File: AdventOfCode/2022/day04.js
const { readLines, part_one, part_two } = require("../../common/aoc");

const pairs = readLines("example").map((a) =>
    a.split(",").map((p) => p.split("-").map((i) => parseInt(i)))
);

function includes([a, b], [c, d]) {
    return (a <= c && b >= d) || (c <= a && d >= b);
}

part_one(pairs.filter(([p1, p2]) => includes(p1, p2)).length);

function overlap([a, b], [c, d]) {
    return a <= d && b >= c;
}

part_two(pairs.filter(([p1, p2]) => overlap(p1, p2)).length);

File: AdventOfCode/2017/day04.py
from libs import *

# Parse input

passphrases = read_lines("example")

cut = passphrases.index("") if read.from_example else len(passphrases)
cut_p2 = cut + 1 if read.from_example else 0

# Part 1


def valid(passphrase):
    words = passphrase.split(" ")
    return len(words) == len(set(words))


part_one(sum([valid(phrase) for phrase in passphrases[:cut]]))

# Part 2


def valid_p2(passphrase):
    words = ["".join(sorted(word)) for word in passphrase.split(" ")]
    return len(words) == len(set(words))


part_two(sum([valid_p2(phrase) for phrase in passphrases[cut_p2:]]))
